WEB-INTEGRATION.md

```diff
@@ -0,0 +1,742 @@
+callback functions and callback interface types in WebIDL.
+
+This document examines various kinds of web platform APIs which accept callbacks
+or otherwise need integration with AsyncContext, and examines whether the
+developer-facing behavior is as one would logically expect. In cases where there
+is a different causally related snapshot, it also considers whether that other
+snapshot should be exposed via another web platform API.
+
+# Individual analysis of web APIs and AsyncContext
+
+Is it really correct to wrap all callbacks which are passed into WebIDL such
+that they restore the AsyncContext snapshot from when they were initially
+provided? Are other snapshots relevant to provide on the side? This section
+examines many web APIs, case by case, and concludes that the initial snapshot is
+appropriate for the invocation of the callback, suggesting certain cases where
+other contexts should also be provided with explicit APIs.
+
+## Web APIs that take callbacks
+
+For web APIs that take callbacks, the context in which the callback is run would
+depend on the kind of API:
+
+### Schedulers
+
+These are web APIs whose sole purpose is to take a callback and schedule it in
+the event loop in some way. The callback will run asynchronously at some point,
+when there is no other JS code in the call stack.
+
+Because there is no other relevant snapshot to consider restoring, the general
+mechanism of saving and restoring the snapshot at the WebIDL callback level is
+correct and sufficient. These APIs should store the context at the time that
+they are called, and invoke the callback with that context.
+
+Examples of scheduler web APIs:
+- [`setTimeout()`](https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#dom-settimeout)
+  [\[HTML\]](https://html.spec.whatwg.org/multipage/)
+- [`setInterval()`](https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#dom-setinterval)
+  [\[HTML\]](https://html.spec.whatwg.org/multipage/)
+- [`queueMicrotask()`](https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#dom-queuemicrotask)
+  [\[HTML\]](https://html.spec.whatwg.org/multipage/)
+- [`requestAnimationFrame()`](https://html.spec.whatwg.org/multipage/imagebitmap-and-animations.html#dom-animationframeprovider-requestanimationframe)
+  [\[HTML\]](https://html.spec.whatwg.org/multipage/)
+- [`requestIdleCallback()`](https://w3c.github.io/requestidlecallback/#dom-window-requestidlecallback)
+  [\[REQUESTIDLECALLBACK\]](https://w3c.github.io/requestidlecallback/)
+- [`scheduler.postTask()`](https://wicg.github.io/scheduling-apis/#dom-scheduler-posttask)
+  [\[SCHEDULING-APIS\]](https://wicg.github.io/scheduling-apis/)
+- [`HTMLVideoElement`](https://html.spec.whatwg.org/multipage/media.html#htmlvideoelement):
+  [`requestVideoFrameCallback()`](https://wicg.github.io/video-rvfc/#dom-htmlvideoelement-requestvideoframecallback)
+  method [\[VIDEO-RVFC\]](https://wicg.github.io/video-rvfc/)
+
+### Async completion callbacks
```

Comment:
Node.js may choose either way--it can decide for itself whether it feels that the alignment with web or ALS is more important for them. I think it'd be good for someone to work on a separate design document for this, but I don't think it's a Stage 2.7-blocker the same way as web integration is.

Reply:
Although this document is specifically about the web integration, because that is a requirement for stage 2.7, I think it should also apply to runtime-specific APIs. Otherwise, that would create an inconsistency in Node.js, since it will have to follow the web API behavior for the web APIs it implements.

I haven't looked at Node.js's APIs in any detail, but if you're talking specifically about async completion callbacks, I don't see how there would be any other possible context for them other than the registration callback. The API (let's say `fs.readFile()`, for a concrete example) would start an async operation in the background, and the data flow of that operation would go back to the call to `fs.readFile()`. I don't think there is any other relevant context here, but @Qard can correct me.

Reply:
I still think Node.js can make this tradeoff for themselves, even as I acknowledge the inconsistency risk.

Reply:
I think the idea was that `fs.readFile` might set some variables internally, and so there's a question of whether those changes would be reflected in the callback.  @Qard seemed to suggest that they should, but this is precisely what leads to the inconsistency between callbacks and promises.

Reply:
This is only observable if `fs.readFile` does in fact set variables internally. As far as I'm aware, this is not the case for any such APIs in Node.js. But you're right that a lot of this doesn't take runtime- (or webplatform-) -internal variables into consideration. And things like the WebIDL handling of callbacks is described as simply storing the snapshot and restoring it later, when the needs of `scheduler.yield` would mean more often than not it's a different snapshot that has to be stored.